File: 215.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        nums[:] = self.divide(nums)
        return nums[::-1][k-1]
    def divide(self, nums):
        length = len(nums)
        if length == 0 or length == 1:
            return nums
        left = self.divide(nums[:length//2] if length%2==0 else nums[:length//2+1])
        right = self.divide(nums[length//2:] if length%2 == 0 else nums[length//2+1:])
        return self.merge(left, right)
    def merge(self, left, right):
        l, r, result = 0, 0, []
        while l<len(left) and r < len(right):
            if left[l] < right[r]:
                result.append(left[l])
                l += 1
            else:
                result.append(right[r])
                r += 1
        return result + left[l:] + right[r:]

File: 215.cpp
class Solution {
public:
    int partition(vector<int>& nums, int l, int r){
        int n = rand()%(r-l+1) + l;
        swap(nums[n], nums[r]);
        int pivot = nums[r];
        int i = l;
        for (int j = l;j<r;j++){
            if (nums[j] < pivot){
                swap(nums[i++], nums[j]);
            }
        }
        swap(nums[i], nums[r]);
        return i;
    }
    void q_sort(vector<int>& nums, int l, int r){
        if (l >= r){
            return;
        }
        int q = partition(nums, l, r);
        q_sort(nums, l, q - 1);
        q_sort(nums, q+1, r);
    }
    
    int findKthLargest(vector<int>& nums, int k) {
        q_sort(nums, 0, nums.size() - 1);
        return nums[nums.size() - k];
    }
};
